### PR TITLE
toplevel: allow multiple phrases in one line

### LINKED ISSUE #8813: Sequencing toplevel phrases in `ocaml`
It seems that `ocaml`'s readline does not allow to sequence toplevel phrases on the same input interaction. Basically the first `;;` seems to act as an "end of line" comment. 

Is there any rationale about this or does it qualify as a bug ? 

More specifically I find the following discrepancy suprising:
```ocaml
> echo 'print_endline "bla";; print_endline "bli";;' > /tmp/bla.ml
> ocaml -noinit
        OCaml version 4.08.0

# #use "/tmp/bla.ml";;
bla
- : unit = ()
bli
- : unit = ()
# print_endline "bla";; print_endline "bli";;
bla
- : unit = ()
# 
```
Also it seems syntax errors are allowed after the first `;;`:
```ocaml
# print_endline "bla";; lat x = 3;;
bla
- : unit = ()
```

### PR DESCRIPTION
**Pull request by damiendoligez**
fixes #8813

As described in https://github.com/ocaml/ocaml/issues/8813#issuecomment-546355599, the toplevel waits for a `;;` **at end of line** (ignoring whitespace) before it starts executing the input.

There are some non-obvious choices to make when handling errors and warnings. You have to choose something that makes sense for the user and is not too hard to implement. This is what I've done:

1. In case of syntax error, the whole line is aborted and nothing gets executed. The error is reported (and the input flushed) as soon as detected, even if there isn't a `;;` at the end of the line.
2. In case of type error or compilation error (or user interrupt while compiling) execution stops at the phrase in error and the rest of the line is ignored.
3. If a phrase raises an exception or the user interrupts its execution, the exception is reported and execution goes on with the rest of the line.

There is a difficulty with displaying warnings and errors. What I've implemented is that errors and warnings display normally for the first phrase, but we switch to batch-mode display for the others.

### CODE REVIEW

File: toplevel/toploop.ml [hunk 1]
```diff
@@ -206,21 +218,42 @@ let loop ppf =
...
+            List.rev (phr :: phrs)
+          else
+            get_phrases (phr :: phrs)
+        | exception Exit -> raise PPerror
+        | exception e -> Location.report_exception ppf e; []
+      in
+      let phrs = get_phrases [] in
+      let saved_report_printer = !Location.report_printer in
+      let nonfirst_report_printer = fun () -> Location.batch_mode_printer in
+      let snap = ref (Btype.snapshot ()) in
```
Comment (gasche) [inline toplevel/toploop.ml]: Unless I am missing something, this line introduces a bug because it shadows the `snap` reference on line 221 which is the one seen by the exception handlers below.

File: toplevel/toploop.ml [hunk 2]
```diff
@@ -206,21 +218,42 @@ let loop ppf =
...
+        let execute i phr =
+          if i > 0 then Location.report_printer := nonfirst_report_printer;
+          snap := Btype.snapshot ();
+          Warnings.reset_fatal ();
+          let phr = preprocess_phrase ppf phr in
+          Env.reset_cache_toplevel ();
+          ignore(execute_phrase true ppf phr)
+        in
+        List.iteri execute phrs
+      end
```
Comment (gasche) [inline toplevel/toploop.ml]: I read the new logic in this function and I believe that it is correct (modulo comment about shadowing `snap`), but it is a beast and not so easy to understand -- I think I would be lost as to what we are doing and why if I wasn't reading the code in the context of the PR itself.

Could you:

1. Move the logic after `get_phrases` to a helper function `execute_phrases`, to make the `loop` function itself shorter and spread the complexity around?
2. add a comment that explains why we change the location printer temporarily?

File: toplevel/toploop.ml [hunk 3]
```diff
@@ -206,21 +218,42 @@ let loop ppf =
...
+      Fun.protect ~finally begin fun () ->
+        let execute i phr =
+          if i > 0 then Location.report_printer := nonfirst_report_printer;
+          snap := Btype.snapshot ();
+          Warnings.reset_fatal ();
+          let phr = preprocess_phrase ppf phr in
+          Env.reset_cache_toplevel ();
+          ignore(execute_phrase true ppf phr)
+        in
+        List.iteri execute phrs
```
Comment (gasche) [inline toplevel/toploop.ml]: Note: another way to write this logic which might be nicer would something like this (sketch):

```ocaml
let rec execute_phrases ~first_phrase = function
| [] -> ()
| phrase :: rest ->
    snap := Btype.snapshot ();
    Warnings.reset_fatal ();
    let phr = preprocess_phrase ppf phr in
    Env.reset_cache_toplevel ();
    ignore (execute_phrase true ppf phr);
    if not first_phrase then execute_phrases ~first_phrase:false rest
    else
      Misc.protect_refs [R Location.(report_printer, batch_printer)] (fun () ->
        execute_phrases ~first_phrase:false rest)
in
execute_phrases ~first_phrase:true phrs
```

**Review by gasche — COMMENTED** [review]
This is a nice touch, thanks! I'm generally in favor (see inline comments).

Could you add a test that exercises this, including in particular the various tweaks (stop-early behavior on error, and the change of printer on errors/warnings)? (I guess this would go in tests/tool-toplevel.)

**Comment by wikku** [comment]
Strictly speaking, it's allowing multiple phrases *per prompt*, the only relevant newline is the one after `;;`. The [manual](https://v2.ocaml.org/manual/toplevel.html) should be updated, though I see that it already incorrectly claims that several phrases are read. I'd remove 

> The system prints a # (sharp) prompt before reading each phrase.

and after that write something like

> Input to the toplevel can span several lines. It begins after the `#` prompt printed by the toplevel and is terminated when the user types `;;` (a double-semicolon) followed by a new line.

File: toplevel/toploop.ml (hunk 1 above)
Comment (damiendoligez) [reply toplevel/toploop.ml]: Good catch, thanks.

File: toplevel/toploop.ml (hunk 2 above)
Comment (damiendoligez) [reply toplevel/toploop.ml]: I've moved the parsing and the processing to their own auxiliary functions and fixed a bug with very long lines that I noticed while doing that.

Comment (damiendoligez) [reply toplevel/toploop.ml]: I've also added a comment to explain the prompt business.

File: toplevel/toploop.ml (hunk 3 above)
Comment (damiendoligez) [reply toplevel/toploop.ml]: Thanks for the suggestion.

**Comment by damiendoligez** [comment]
@wiktorkuchta thanks for the suggestion. I've followed it and also fixed the description for multiple definitions.

I'll note that there is still something slightly annoying: a comment after the `;;` triggers the "ask for more input" behavior. I don't see how to avoid that without writing yet another comment parser, so I won't fix it. Fortunately, there is an easy workaround: type a `;;` after the comment (or on the next line).

File: Changes [hunk 4]
```diff
@@ -487,6 +487,9 @@ Working version
 
 ### Bug fixes:
 
+- #8813, #12029: In the toplevel, let the user type several phrases in one line
+  (Damien Doligez, report by Daniel Bünzli, review by Gabriel Scherer)
```
Comment (gasche) [inline Changes]: and Wiktor Kuchta

File: man/ocaml.1 [hunk 5]
```diff
@@ -37,23 +37,19 @@ command is the toplevel system for OCaml,
 that permits interactive use of the OCaml system through a
 read-eval-print loop. In this mode, the system repeatedly reads OCaml
 phrases from the input, then typechecks, compiles and evaluates
-them, then prints the inferred type and result value, if any. The
+them, then prints the inferred type and result value, if any.
 system prints a # (hash) prompt before reading each phrase.
```
Comment (gasche) [inline man/ocaml.1]: This diff looks weird, is there an issue here? (`The` is gone?)

File: man/ocaml.1 [hunk 6]
```diff
@@ -37,23 +37,19 @@ command is the toplevel system for OCaml,
...
+(sharp) prompt printed by the system and is terminated by ;; (a
+double-semicolon) followed by optional white space and an end of line.
+The toplevel input consists in one or several toplevel phrases.
 
-The toplevel system is started by the command
-.BR ocaml (1).
-Phrases are read on standard input, results are printed on standard
-output, errors on standard error. End-of-file on standard input
-terminates
-.BR ocaml (1).
```
Comment (gasche) [inline man/ocaml.1:51]: Why is this removed, is it redundant with some other place?

File: toplevel/toploop.ml [hunk 7]
```diff
@@ -184,6 +184,54 @@ let load_ocamlinit ppf =
...
+      can't move the cursor back to the correct line
+ *)
+let process_phrases ppf snap phrs =
+  match phrs with
+  | [] -> ()
+  | phr :: rest ->
+    process_phrase ppf snap phr;
+    Misc.protect_refs
+      Location.[R (report_printer, fun () -> batch_mode_printer)]
+      (fun () -> List.iter (process_phrase ppf snap) rest)
```
Comment (gasche) [inline toplevel/toploop.ml]: Nice. (We could use `if rest <> [] then Misc.protect_refs ...` but there is no strong need.)

File: driver/main_args.ml [hunk 8]
```diff
@@ -336,6 +336,9 @@ let mk_noinit f =
 let mk_nolabels f =
   "-nolabels", Arg.Unit f, " Ignore non-optional labels in types"
 
+let mk_prompt f =
+  "-prompt", Arg.Unit f, " Output prompts (default)"
```
Comment (gasche) [inline driver/main_args.ml:345]: What is the relation between this change and the PR?

File: manual/src/cmds/top.etex [hunk 9]
```diff
@@ -26,10 +27,10 @@ directive-argument:
         | 'true' || 'false'
 \end{syntax}
 
-A phrase can consist of a definition, like those found in
+A phrase can consist of a sequence of definitions, like those found in
```
Comment (gasche) [inline manual/src/cmds/top.etex:30]: Do I correctly understand that this change is correct but unrelated to the present PR? (The PR implements "several phrases per prompt", and here you add an independent remark that `let x = 1 let y = 2` counts as a single phrase?)

**Review by gasche — COMMENTED** [review]
The new implementation is fine. I wonder why we have a new `-prompt` option? And the manual changes were a bit harder to proofread than I would have liked because sometimes I don't know what their purpose is.

File: man/ocaml.1 (hunk 5 above)
Comment (damiendoligez) [reply man/ocaml.1]: In fact I meant to remove the whole sentence.

File: man/ocaml.1 (hunk 6 above)
Comment (damiendoligez) [reply man/ocaml.1:51]: It's both wrong (everything is printed on stdout) and mostly redundant with the above, except for the bit about end-of-file. I've added it back.

File: driver/main_args.ml (hunk 8 above)
Comment (damiendoligez) [reply driver/main_args.ml:345]: This change is needed for the tests: ocamltest passes `-noprompt` to the toplevel and we need a way to get the prompts back (otherwise we can't tell when phrases have been processed together).

File: manual/src/cmds/top.etex (hunk 9 above)
Comment (damiendoligez) [reply manual/src/cmds/top.etex:30]: This is correct. Same for the change to the grammar fragment, where the sequence of definitions can be empty.

File: toplevel/toploop.ml (hunk 7 above)
Comment (damiendoligez) [reply toplevel/toploop.ml]: I've followed your suggestion.

File: testsuite/tests/typing-extensions/cast.ml [hunk 10]
```diff
@@ -94,10 +94,10 @@ let c3: castable = pop_castable ();;
 
 (* We can also downcast these values to foos and bars *)
 
-let f1: foo = c1#cast (Class Foo);; (* Ok *)
-let f2: foo = c2#cast (Class Foo);; (* Ok *)
-let f3: foo = c3#cast (Class Foo);; (* Ok *)
+let f1: foo = c1#cast (Class Foo) (* Ok *);;
+let f2: foo = c2#cast (Class Foo) (* Ok *);;
+let f3: foo = c3#cast (Class Foo) (* Ok *);;
```
Comment (gasche) [inline testsuite/tests/typing-extensions/cast.ml]: What happens if you leave the `;;` in their previous place there? (Does this change has a backward-compatibility impact we should worry about?)

Comment (damiendoligez) [reply testsuite/tests/typing-extensions/cast.ml]: If you don't move the `;;` (or at least add another one at the end of the line) the test will fail because the toplevel is waiting for more input. I'm not too worried about the backward-compatibility impact, unless some users have been using the toplevel for batch processing. For interactive use, I don't think many people type comments, and the ones who do will notice the problem immediately and type `;;`.

Comment (gasche) [reply testsuite/tests/typing-extensions/cast.ml]: Concretely if I have a `test.ml` file with contents

```
let x = 1;; (* one *)
let y = 2;; (* two *)
```

then both `ocaml test.ml` and `cat test.ml | ocaml` will process both phrases today. (The former in a silent way that is equivalent to `cat test.ml | ocaml -stdin` I believe.)  Has this changed in any way?

Comment (damiendoligez) [reply testsuite/tests/typing-extensions/cast.ml]: Yes, and I believe this will only impact the testsuite of OCaml itself.

Comment (gasche) [reply testsuite/tests/typing-extensions/cast.ml]: Does your "yes" mean that the behavior have changed? If so, how did it change? (Is the last phrase skipped?)

Comment (damiendoligez) [reply testsuite/tests/typing-extensions/cast.ml]: Yes, the behavior has changed. The prompts won't be in the same places, and the last group of phrases will not be executed.

last group of phrases: everything between the last double-semicolon-at-end-of-line and the end of the file.

I'm now thinking of an alternate implementation that would let us properly ignore comments and solve this problem.

**Comment by damiendoligez** [comment]
@gasche I've fixed the problem with comments and removed the changes to the test files. I've also added a test for the feature.

**Comment by gasche** [comment]
@damiendoligez fixed the problem again with a new, better approach (again!) that we discussed around a coffee yesterday

**Comment by gasche** [comment]
What is going on here is fairly subtle, and I still don't find the code comments so clear. What I propose to do is to explain the overall system as I understand it (in which the changes are taking place), explain what is going on in the change, and then approve the proposed implementation which I believe is correct -- it is an improvement to the user experience at the cost of an extra layer of complexity.

There are several interacting systems in the toplevel 
What is going on with the change:

- There is an I/O layer that is in charge of reading input from the user
  *and* emitting prompts to clarify that it is (still) waiting for input.

  This layer must be careful to read user input character-by-character,
  to avoid blocking reads, and to not read past the "last" newline
  that the user sent to the toplevel.

- There is a standard lexer and parser that consumes the input buffer from the I/O layer and turns it into OCaml AST.

- Finally the parsed AST is sent to an evaluator which does something the user can see, and then we forget about the current user input and ask for new input.

The information is not only flowing "down" in this diagram: when the toplevel code calls the parser, it calls the lexer, which is itself in charge of asking the I/O layer for more input (by calling the "refill my lexing buffer" function, which... has the observable effect of producing a multiline-prompt in some cases).

In particular, the I/O layer will stop on the first newline it encounters at first (it has no domain knowledge of whether its input is complete), and will only read multiline input if the lexer (called by the parser) asks for more to parse a complete phrase.

Now, let's describe the change:

- Before, we would call the parser on the user input provided by the I/O layer, and stop at the first sentence parsed. This has the effect of ignoring all the input after the first ";;" -- it is seen by the I/O layer but never by the parser.
- This PR adds a new "input driver" layer *on top* of the parser, which repeatedly calls the parser to parse *several* phrases, until a certain stopping condition is met.

  The condition implemented is as follows: we stop after parsing a phrase if only whitespace remains between the end of the phrase and the end of the line.

  The meaning of "whitespace" here is "things that the standard OCaml lexer discards instead of returning a token", it also includes *complete* comments.

It is a bit tricky to implement correctly this condition of "only whitespace including comments". The way @damiendoligez does it is to copy the prospective "end of input" part, and call the lexer on that. If it neither returns a token nor complain about an unterminated comment, then we are good. (It if complains about an unterminated string literal, then this is not the end of input.)

**Review by gasche — APPROVED** [review]

**Comment by gasche** [comment]
Note: @dbuenzli's `down.top` overlay implements its own I/O layer, so if you want to experiment with what is going on here (or benefit from the subtle prompt displayed when in the middle of an unterminated comment) you should disable it.

**Comment by damiendoligez** [comment]
@gasche you may want to review the latest commit: one more hack to get the right continuation prompt when we refill the buffer because of an unterminated comment.

**Comment by gasche** [comment]
The last commit made me realize that your code is doing something potentially dangerous: we are calling the `Lexer.token` function on different source buffers, and we expect the lexings to be independent from each other. This is in fact wrong because the lexer maintains global mutable state across calls. Most of this global mutable state is reset to a default value when `token` returns, so this is okay, but not all of it, in particular there is a global table of comments, we will duplicate comments if we lex the "shadow" input on the side.

I have an idea to fix this -- to have local state in the lexer -- by putting the lexer code inside a functor, I may submit a separate PR to do this.

Then, if we have local state: we could change the lexer code to not reset the `is_in_comment` boolean when raising an Unterminated_comment exception, and then your workaround would not be necessary anymore.

**Comment by damiendoligez** [comment]
That's a good remark, I didn't think of side effects in the lexer. There's one you missed, which means I have one more bug to fix: the lexer can output warnings, and I need to disable them when lexing the shadow input.

So don't merge yet.

**Comment by damiendoligez** [comment]
Oh, and as far as comments are concerned, the toplevel doesn't do anything with them, so we don't have to fix that part.

**Comment by gasche** [comment]
> Oh, and as far as comments are concerned, the toplevel doesn't do anything with them, so we don't have to fix that part.

Your approach is "thinking about non-temporary state changes one by one and claiming that it is okay to ignore them for reasons". That sounds like a sure way to shoot ourselves in the foot, either today (because we missed something) or tomorrow (because we add more state / change the way state works).

**Comment by damiendoligez** [comment]
That's right but, as you said, adding local state to the lexer is best left to another PR, which we don't have to do urgently because in this case it's OK to duplicate comments.

**Comment by damiendoligez** [comment]
@gasche Do you want to review the latest commit ?

**Review by gasche — APPROVED** [review]
Thanks for the ping. I believe that the last commit is correct, and approve the PR.
(I remain surprised at the amount of complexity involved in this feature and the toplevel parsing code in general. Parsing interactive user input is no joke.)

**Comment by gasche** [comment]
@damiendoligez it looks like you need to do something, the CI for this PR seems to fail. I observed a flay failure of `intext_par.ml` on macos, but there are also failures related to the PR.

**Comment by gasche** [comment]
Gentle ping: this PR feels almost-ready and an important amount of work. We should either give up and close it now, or try to get it merge without waiting too long to avoid the bitrot.
